src/sinks/clickhouse/parse.rs

```diff
@@ -0,0 +1,219 @@
+use clickhouse_rs::types::{SqlType, DateTimeType};
+use chrono_tz::Tz;
+use nom::{
+    bytes::complete::{tag, take_until1},
+    combinator::{all_consuming, map_res},
+    character::complete::{u8 as p_u8, u32 as p_u32, u64 as p_u64},
+    sequence::{delimited, preceded, pair},
+    branch::alt,
+    IResult
+};
+
+macro_rules! parse_static {
+    ($($lit: literal, $method: ident, $id: ident),*) => {
+        $(
+            fn $method(input: &str) -> IResult<&str, SqlType> {
+                let (rest, _) = tag($lit)(input)?;
+                Ok((rest, SqlType::$id))
+            }
+        )*
+    }
+}
+
+parse_static!{
+    "UInt8", parse_uint8, UInt8,
+    "UInt16", parse_uint16, UInt16,
+    "UInt32", parse_uint32, UInt32,
+    "UInt64", parse_uint64, UInt64,
+    "Int8", parse_int8, Int8,
+    "Int16", parse_int16, Int16,
+    "Int32", parse_int32, Int32,
+    "Int64", parse_int64, Int64,
+    "Float32", parse_float32, Float32,
+    "Float64", parse_float64, Float64,
+    "String", parse_string, String,
+    "UUID", parse_uuid, Uuid,
+    "Date", parse_date, Date,
+    "IPv4", parse_ipv4, Ipv4,
+    "IPv6", parse_ipv6, Ipv6
+}
+
+pub(super) fn parse_field_type(s: &str) -> IResult<&str, SqlType> {
+    all_consuming(parse_sql_type)(s)
+}
+
+fn parse_sql_type(s: &str) -> IResult<&str, SqlType> {
+    alt((
+        // types that can be wrapped by Nullable(xxx)
+        parse_nullable_inner,
+        // Nullable
+        parse_nullable,
+        // types that can NOT be wrapped by Nullable
+        parse_array,
+        parse_map,
+    ))(s)
+}
+
+fn parse_static_type(s: &str) -> IResult<&str, SqlType> {
+    alt((
+        parse_uint8,
+        parse_uint16,
+        parse_uint32,
+        parse_uint64,
+        parse_int8,
+        parse_int16,
+        parse_int32,
+        parse_int64,
+        parse_float32,
+        parse_float64,
+        parse_string,
+        parse_uuid,
+        parse_date,
+        parse_ipv4,
+        parse_ipv6,
+    ))(s)
+}
+
+
+fn parse_array(s: &str) -> IResult<&str, SqlType> {
+    let (rest, v) = preceded(
+        tag("Array"), 
+        delimited(tag("("), parse_sql_type, tag(")"))
+    )(s)?;
+    Ok((rest, SqlType::Array(v.into())))
```

Comment:
I think the `map` form above works for all of these:
```suggestion
    preceded(
        tag("Array"), 
        delimited(tag("("), parse_sql_type, tag(")"))
    )(s)
    .map(|(rest, v)| (rest, SqlType::Array(v.into())))
```